### ISSUE TITLE
libgirepository-1.0.so.1 error from briefcase dev: Undefined symbol g_string_free_and_clear

### ISSUE THREAD

**Issue by rasmusJenle** [beeware/briefcase#1580]
### Describe the bug

I am quite sure that what is described below is not a bug, but simply a problem I cannot solve myself. When I posted the problem on SO, I was recommended to ask my question here.

I am trying to follow the Beeware Tutorial. When I get to the point of running Briefcase (i.e. <$ briefcase dev>) in part 1 I get the error:

ImportError: /lib/x86_64-linux-gnu/libgirepository-1.0.so.1: undefined symbol: g_string_free_and_steal

Having looked in various forums (e.g. https://www.reddit.com/r/Fedora/comments/12z32r1/davinci_resolve_libpango_undefined_symbol_g/), I have become fairly convinced that I am encountering library compatibility issues caused by a mismatch between versions.

The closest I have come to a solution involves downgrading to Python 3.7 (using miniconda3 to create the environment), while following the (remaining) instructions in the BeeWare Tutorial. This approach enables me to run <$ briefcase dev> and open the empty app in developer mode a first time. However, the second time I run <$ briefcase dev> within the same setup, I am met with the import error again.

How can I set up an environment that supports Briefcase?

### Steps to reproduce

$ sudo apt update sudo apt install build-essential git pkg-config python3-dev python3-venv libgirepository1.0-dev libcairo2-dev gir1.2-webkit2-4.0 libcanberra-gtk3-module

...

$ mkdir beeware-tutorial
cd beeware-tutorial
python3 -m venv beeware-venv
source beeware-venv/bin/activate

...

$ python -m pip install briefcase

...

$ briefcase new

```

$ briefcase dev

[helloworld] Installing requirements...
Collecting toga-gtk~=0.4.0
  Obtaining dependency information for toga-gtk~=0.4.0 from https://files.pythonhosted.org/packages/5e/c4/77db3896dcef6c4ec73bbcf827183c6659cb49b63f646c9c400191f4e1ff/toga_gtk-0.4.0-py3-none-any.whl.metadata
  Using cached toga_gtk-0.4.0-py3-none-any.whl.metadata (3.0 kB)
Collecting pytest
  Obtaining dependency information for pytest from https://files.pythonhosted.org/packages/f3/8c/f16efd81ca8e293b2cc78f111190a79ee539d0d5d36ccd49975cb3beac60/pytest-7.4.3-py3-none-any.whl.metadata
  Using cached pytest-7.4.3-py3-none-any.whl.metadata (7.9 kB)
Collecting toga-core==0.4.0 (from toga-gtk~=0.4.0)
  Obtaining dependency information for toga-core==0.4.0 from https://files.pythonhosted.org/packages/c5/63/108254beaa0f14aae11af5ab78be4c1dbb3dbd9e94ebc7202a8eb7a01a0e/toga_core-0.4.0-py3-none-any.whl.metadata
  Using cached toga_core-0.4.0-py3-none-any.whl.metadata (4.9 kB)
Collecting gbulb>=0.5.3 (from toga-gtk~=0.4.0)
  Using cached gbulb-0.6.4-py3-none-any.whl (17 kB)
Collecting pycairo>=1.17.0 (from toga-gtk~=0.4.0)
  Using cached pycairo-1.25.1-cp311-cp311-linux_x86_64.whl
Collecting pygobject>=3.46.0 (from toga-gtk~=0.4.0)
  Using cached PyGObject-3.46.0-cp311-cp311-linux_x86_64.whl
Collecting travertino>=0.3.0 (from toga-core==0.4.0->toga-gtk~=0.4.0)
  Obtaining dependency information for travertino>=0.3.0 from https://files.pythonhosted.org/packages/25/bf/59fd5d89c3b0c5d897862ac83457fe82d934dde52d59a9106df6066e97b2/travertino-0.3.0-py3-none-any.whl.metadata
  Using cached travertino-0.3.0-py3-none-any.whl.metadata (4.3 kB)
Collecting iniconfig (from pytest)
  Using cached iniconfig-2.0.0-py3-none-any.whl (5.9 kB)
Requirement already satisfied: packaging in /home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages (from pytest) (23.2)
Collecting pluggy<2.0,>=0.12 (from pytest)
  Obtaining dependency information for pluggy<2.0,>=0.12 from https://files.pythonhosted.org/packages/05/b8/42ed91898d4784546c5f06c60506400548db3f7a4b3fb441cba4e5c17952/pluggy-1.3.0-py3-none-any.whl.metadata
  Using cached pluggy-1.3.0-py3-none-any.whl.metadata (4.3 kB)
Using cached toga_gtk-0.4.0-py3-none-any.whl (50 kB)
Using cached toga_core-0.4.0-py3-none-any.whl (521 kB)
Using cached pytest-7.4.3-py3-none-any.whl (325 kB)
Using cached pluggy-1.3.0-py3-none-any.whl (18 kB)
Using cached travertino-0.3.0-py3-none-any.whl (16 kB)
Installing collected packages: travertino, pycairo, pluggy, iniconfig, toga-core, pytest, pygobject, gbulb, toga-gtk
Successfully installed gbulb-0.6.4 iniconfig-2.0.0 pluggy-1.3.0 pycairo-1.25.1 pygobject-3.46.0 pytest-7.4.3 toga-core-0.4.0 toga-gtk-0.4.0 travertino-0.3.0

[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: pip install --upgrade pip
Installing dev requirements... done

[helloworld] Starting in dev mode...
===========================================================================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen runpy>", line 226, in run_module
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/home/rasmus_jenle/projects/beeware-tutorial/helloworld/src/helloworld/__main__.py", line 4, in <module>
    main().main_loop()
    ^^^^^^
  File "/home/rasmus_jenle/projects/beeware-tutorial/helloworld/src/helloworld/app.py", line 27, in main
    return HelloWorld()
           ^^^^^^^^^^^^
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/toga/app.py", line 398, in __init__
    self.factory = get_platform_factory()
                   ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/toga/platform.py", line 114, in get_platform_factory
    factory = importlib.import_module(f"{backend.value}.factory")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rasmus_jenle/miniconda3/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/toga_gtk/factory.py", line 1, in <module>
    from . import dialogs
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/toga_gtk/dialogs.py", line 4, in <module>
    from .libs import Gtk
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/toga_gtk/libs/__init__.py", line 2, in <module>
    from .gtk import *  # noqa: F401, F403
    ^^^^^^^^^^^^^^^^^^
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/toga_gtk/libs/gtk.py", line 1, in <module>
    import gi
  File "/home/rasmus_jenle/projects/beeware-tutorial/beeware-venv/lib/python3.11/site-packages/gi/__init__.py", line 40, in <module>
    from . import _gi
ImportError: /lib/x86_64-linux-gnu/libgirepository-1.0.so.1: undefined symbol: g_string_free_and_steal

Problem running app helloworld.

```

### Expected behavior

Expected to start the BeeWare app in developer mode.

### Environment

- Operating System: Ubuntu 23.10
- Python version: 3.11
- Software versions:
  - Briefcase: 0.3.16

### Logs

```

```

[briefcase.2023_12_18-20_22_52.dev.log](https://github.com/beeware/briefcase/files/13707804/briefcase.2023_12_18-20_22_52.dev.log)

**Comment by rmartin16** [comment]
I upgraded my instance of 23.04 to 23.10 but cannot replicate this issue.

The symbol `g_string_free_and_steal` should be found via `libglib-2.0.so.0`. The fact it is not suggests something might be "wrong" with your system.

What is the output for `ldd /lib/x86_64-linux-gnu/libgirepository-1.0.so.1`?

**Comment by rasmusJenle** [comment]
Thank you for your response. I get the following:

$ ldd /lib/x86_64-linux-gnu/libgirepository-1.0.so.1
	linux-vdso.so.1 (0x00007ffff3fce000)
	libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007feb17474000)
	libgobject-2.0.so.0 => /lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00007feb17413000)
	libgmodule-2.0.so.0 => /lib/x86_64-linux-gnu/libgmodule-2.0.so.0 (0x00007feb1740c000)
	libgio-2.0.so.0 => /lib/x86_64-linux-gnu/libgio-2.0.so.0 (0x00007feb17238000)
	libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007feb1722d000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007feb17140000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007feb16e00000)
	libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007feb170a5000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007feb17086000)
	libmount.so.1 => /lib/x86_64-linux-gnu/libmount.so.1 (0x00007feb1703a000)
	libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007feb1700d000)
	/lib64/ld-linux-x86-64.so.2 (0x00007feb175f2000)
	libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1 (0x00007feb16dc3000)

**Comment by rmartin16** [comment]
Thanks. As well, what is the output for `ldd --version` and `file /lib/x86_64-linux-gnu/libglib-2.0.so.0`?

**Comment by rasmusJenle** [comment]
For those commands I get:

$ ldd --version
ldd (Ubuntu GLIBC 2.38-1ubuntu6) 2.38
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

$ file /lib/x86_64-linux-gnu/libglib-2.0.so.0
/lib/x86_64-linux-gnu/libglib-2.0.so.0: symbolic link to libglib-2.0.so.0.7800.0

**Comment by rmartin16** [comment]
Thanks. So, that version of `glib` is new enough (unless the filename is misleading) to contain `g_string_free_and_steal` (since it was introduced in 7600). I was expecting this to find an older version of `glib` given the issue description.

At this point, it seems something may be different about finding `glib` through the process of `briefcase dev`; with that in mind, can you please upload the output of this command:
```
LD_DEBUG=all briefcase dev 2>&1 | grep -B10 -A5 g_string_free_and_steal
```
Using `LD_DEBUG` produces a tremendous amount of output...so, I'm trying to trim it down to just the interesting bits. Because of that, though, be patient if it doesn't appear to be doing anything at first.

**Comment by rasmusJenle** [comment]
I appreciate the fact that you warned me about having to wait for LD_DEBUG to produce a result. However, it doesn't output anything to the terminal:

(base) rasmus_jenle@25711131719:~/projects$ LD_DEBUG=all briefcase dev 2>&1 | grep -B10 -A5 g_string_free_and_steal
(base) rasmus_jenle@25711131719:~/projects$

**Comment by rmartin16** [comment]
That may be because it can't find the `briefcase` command. Or at least, the `beeware-venv` virtual environment doesn't appear to be activated; I think it's using the base Conda environment. This is also likely true if the command ran quite quickly and nothing was printed to the console.

Can you try re-running that command after activating the Python virtual environment with this command:
```
source beeware-venv/bin/activate
```
You'll need to be in the `beeware-tutorial` directory where the `beeware-venv` directory should exist.

**Comment by rasmusJenle** [comment]
My apologies. It got late by the end. Regretfully, running LD_DEBUG from `beeware-venv` provides the same result. Like when running it in `base`, it also finishes immediately.
```
(base) rasmus_jenle@25711131719:~/projects/beeware-tutorial$ source beeware-venv/bin/activate
(beeware-venv) (base) rasmus_jenle@25711131719:~/projects/beeware-tutorial$ LD_DEBUG=all briefcase dev 2>&1 | grep -B10 -A5 g_string_free_and_steal
(beeware-venv) (base) rasmus_jenle@25711131719:~/projects/beeware-tutorial$ 
```
Side note: It strikes me as strange that I am having all these problems now, since I've just started from a blank with my Ubuntu install. I don't think I've even had the chance to fumble around and ruin stuff yet.

**Comment by rasmusJenle** [comment]
Due to your sobering remarks, I have now found out what the problem is. I created the `venv` from within anaconda's `base`, as opposed to deactivating anaconda's `base` and subsequently creating the `venv`. Having started over by following the BeeWare Tutorial after deactivating `base`, I can now run `$ briefcase dev` repeatedly without problems.

Thank you so much for your time. I really look forward to trying out BeeWare.
